packages/fxa-auth-server/scripts/audit-orphaned-customers.ts

```diff
@@ -0,0 +1,319 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import program from 'commander';
+import Stripe from 'stripe';
+import Container from 'typedi';
+import { setupProcessingTaskObjects } from '../lib/payments/processing-tasks-setup';
+import { promises as fs } from 'fs';
+import {
+  AppConfig,
+  AuthLogger
+} from '../lib/types';
+import AppError, { ERRNO } from '../lib/error';
+
+type CustomerData = {
+  id: string;
+  uid?: string;
+  email?: string;
+};
+
+type Customers = Map<string, CustomerData>
+
+const pckg = require('../package.json');
+
+const isOfStatus = (
+  keyInput: string
+): keyInput is Stripe.Subscription.Status => {
+  return [
+    'active',
+    'canceled',
+    'incomplete',
+    'incomplete_expired',
+    'past_due',
+    'trialing',
+    'unpaid',
+  ].includes(keyInput);
+};
+
+const isCustomer = (
+  customer: string | Stripe.Customer | Stripe.DeletedCustomer
+): customer is Stripe.Customer => {
+  if (typeof customer !== 'string') {
+    return !customer.deleted;
+  }
+  return false;
+};
+
+const parseDateForFirestore = (date: number) => Math.floor(date / 1000);
+
+const parseStartDate = (date: string): number =>
+  parseDateForFirestore(
+    date ? Date.parse(date) : Date.parse('2021-08-13T00:00:00')
+  );
+
+const parseEndDate = (date: string): number =>
+  parseDateForFirestore(
+    date ? Date.parse(date) : Date.parse('2023-02-13T23:59:59')
+  );
+
+const parseSubscriptionStatus = (
+  status: string
+): Stripe.Subscription['status'] | undefined => {
+  if (status) {
+    if (!isOfStatus(status)) {
+      throw new Error(
+        `Please provide a valid status. Invalid status: ${status}`
+      );
+    }
+    return status as Stripe.Subscription['status'];
+  }
+  return undefined;
+};
+
+const parseLogCadence = (cadence: number): number => cadence || 500;
+
+function convertToCsv(data: CustomerData[]) {
+  const replacer = (key: string, value: any) => (value === null ? '' : value); // specify how you want to handle null values here
+  const header = Object.keys(data[0]);
+  return [
+    header.join(','), // header row first
+    ...data.map((row: CustomerData) =>
+      header
+        .map((fieldName: string) =>
+          JSON.stringify(row[fieldName as keyof typeof row], replacer)
+        )
+        .join(',')
+    ),
+  ].join('\r\n');
+}
+
+async function writeCsv(data: CustomerData[], file: string) {
+  if (!data.length) {
+    return;
+  }
+  const csv = convertToCsv(data);
+  try {
+    await fs.writeFile(`${parseDateForFirestore(Date.now())}_${file}.csv`, csv);
+  } catch (err) {
+    console.error(err);
+  }
+}
+
+const orphanedCustomers: CustomerData[] = [];
```

Comment:
Just a comment, no change necessary. When possible, rather don't use global variables. Can sometimes be tricky to keep track of when and where they are changed and used. In this script it's fine though.